### PR TITLE
Release v4.0.0-rc.1

### PR DESCRIPTION
**Pull request by nodech**
Check [Release Notes Draft](https://github.com/handshake-org/hsd/blob/8047df7130af7c00ba49d4fe7aac9ec797b23616/docs/release-notes/release-notes-v4.x-draft.md) (Draft because it's release candidate)

### CODE REVIEW

**Comment by befranz** [comment]
Thinking about to adopt my install/update script for it. All the full nodes installed with the script are hsd v3.01.
Do I get it right that I pass all options ```--compact-tree-on-init```, ```—compact-tree-init-interval``` and probably ```--agent``` as command parameters and not as an option in the config file hsd.conf?
Do I have to remove ```—chain-migrate=3```and ```--compact-tree-on-init``` after the first initial start of hsd v4?
And can I leave ```—compact-tree-init-interval``` as default for every start?

**Comment by nodech** [comment]
> Thinking about to adopt my install/update script for it. All the full nodes installed with the script are hsd v3.01. Do I get it right that I pass all options `--compact-tree-on-init`, `--compact-tree-init-interval` and probably `--agent` as command parameters and not as an option in the config file hsd.conf? Do I have to remove `--chain-migrate=3`and `--compact-tree-on-init` after the first initial start of hsd v4? And can I leave `--compact-tree-init-interval` as default for every start?

Thanks for the feedback. I guess these features need more description.
I added more details to these features.

  Let's start with update of the scripts. This is Release Candidate (beta) and we will take some time (around a week maybe?) after we release and get feedback from the users who use beta. If we don't have any issues we will release `v4.0.0`. I believe it would be better to update installation script after `v4.0.0` is out of beta.

  `--agent` is highly unlikely that users will need to use. It is mostly meant to signal custom clients that have either modified, embeded or extended `hsd` in some way but are part of the network and want to signal it. Most of the time even these nodes may not want to signal it. For example, Bob wallet could theoritically have listening nodes (which I believe they don't) - In that case they could signal `/hsd:version/bob:version/` - But in case there's issue with bob side software they can easily targeted and we are talking about "normal users" and not "super users" who can do bare necessary to secure their nodes. So I would say installer that is meant to be for everybody, does not need special agent.
  
  `--compact-tree-on-init` and `--compact-tree-init-interval` are similar to the `prune` option. So I would say, it depends on the configuration you are going for. For example, if you want Full Node in its truest sense - which needs to have fulll history of the tree and be able to provide proofs for the tree root in the history - then you don't use these options. Similar to how you may not use `prune`. If you are going for "disk space efficiency" - and take every optimization you can get from hsd side then you will enable both: `--compact-tree-on-init` and `--prune`.  
  I tried to expand description of these two options in the release notes, you can check updated docs.
  
  For the question of `hsd.conf` - anything that can be passed as a shell argument to the `hsd` can be provided in `hsd.conf`. `compact-tree-on-init`, `compact-tree-init-interval` and `agent` are all the options that will go to the `hsd.conf`.
  
  I believe there was some doc describing how it worked, but I will just try to explain here anyway. When we have database changes in the codebase (as in v4), node will not start if the migration has not happened. In general this flag is not necessary, this is only used to tell the hsd to run the migrations that are pending and you provide the latest migration ID in that. If the ID is not the latest, hsd will complain. For example users who upgraded from `v2` to `v3` may already have`chain-migrate: 2` in their configuration file, they can just bump that version. If you don't want to have it in the `hsd.conf`, you can start node with `--chain-migrate=3` once and let it run or even restart after it's done w/o argument.
  So basically, that flag is only necessary to be passed when node has not done migration. That's the way we tell users to take action and also a way for users to reconsider if you want to upgrade now. After migration is done, you can either remove that flag or leave it. It wont complain if you pass `--chain-migrate=3` and it has already done migratoins.  
  Note fresh nodes don't need to migrate, they are already using latest database scheme.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "hsd",
-  "version": "4.99.0",
+  "version": "4.00.0-rc.1",
```
Comment (pinheadmz) [inline package.json:3]: does this middle number need to be single `0`? I dont care just wondering

**Review by pinheadmz — APPROVED** [review]
ACK ae9edbf6318580a70a775bd250ea1975ca87fe1e
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK ae9edbf6318580a70a775bd250ea1975ca87fe1e
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmKkre0ACgkQ5+KYS2KJ
yTpl/Q//TILGE1X88XxFDRfqQsK0uzFzj5HGJXBszrkLwgb8tkol1exkt4iToCA3
RaN1SzWsCPaSwZ60zzvkKOEAPgc0jYko7MZt9pgOGYiF+9eKGEyPitABDWhvF8Hb
cVGAHm8uyyN08khnSDXI2jrsn1Z7JaUP6KNf8Igb647PTqvfVmdEkXSzcBIXnZPd
r2RKuDLeMJKcL3Uw3TLIAEVMjk3CJpaUi6BRwPA25hvxoGvbLoBNf6PrdTfL26Bn
BCFUfw89Icwo50L/OPIix7WkuEzIUy7IokY7kyJh88mn5Wp+rQDEb1f33sDiXkrg
U2xv690DKgGAesIzRboBcpygdGA7tS2yQn0n2fDdwg5EzJjt0EAIAXnoVcJX9HVo
rk69+oHfPDC2KMs63bhm2UPgn9/gtMdtv8MV48ZZZ9w7M4R+q2NYSfu85MgiCKgb
QMmHqjW3tQdGSdeCr3wWvGJmegBgrLuqHQ7DJ8lBGOk9hHQso+C9UC9F3ANsENvA
Z8qw9zXOq9dD2CTgcdKHJcvhGczuIpZAdB2hJB5VhyVMuXjMzlJn82snENxFTAmh
/FyYMJcg4TdJMblHAH3rDd+qxnTKQ9npiU/mIqgHr0TsyY2jxxIG6KEvLAYqQKjS
gYqqWMacGbi3PZFn7TmBE5iXcS773G2EbErp/T5ar6JyiZcPGU4=
=moLh
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

File: package.json (hunk 1 above)
Comment (nodech) [reply package.json:3]: yep, it should. typo

**Review by pinheadmz — APPROVED** [review]
ACK 76a7eea11b2bab3e283d8cf25181076db5376e97
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 76a7eea11b2bab3e283d8cf25181076db5376e97
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmKkr0QACgkQ5+KYS2KJ
yTpboxAAizxA/ATXCkT1ho8X6SV3vrUnFLjn/E1Y6SGVEIcfqFBngih5Z8La/I1R
Q8AoCKLplh2ow1seMy9ekUPkH/wSi2587hK08KB8+HzKvUHuvicwSzd6IVnURKmN
AnpuHAzAMemH+qFBi1oUPi8+Cx0jYVBSC0HNf2ODhJXJOdTOrCL7/sMqggzHc3vo
MroJCa9n2aNsBNVG9nuS8oaV//7V6wAyYk7tQ1D5bush5qqrEEawphKh4pRR+MaP
mdH88hN0AmyqLy8kvgPFnFw4LelusX7FPCV+/EA98r39rvAPd/4/FFaOcj3KCSHI
ZBURI1WnYWxv6xJYSWSQ6ilCZR5xZqWmeDFiPKjZdMqpX7aRQn2pNa+aAVhb3CA9
TTAbrpWCur1DpkCXWRvGgptYaR5C/NoZ+RMigvGpSKsaKuMcWkzIZ4x8K1t4jwhV
ogDtV1Gd/cMpcyCRyUaIOgkxD6PBFcN47oq9rBore4fSsuQLc4dj/7cnEm+KMP0J
LY1V55K5jd4t0/Zg1wpHydCfwiVhYuQE0nG8zUTVzQy1L7G9OvnfOdfoMtvSNCTF
73uVYSXgbGfnbTKohGeTiB3O5QhIOmKg8b39WYNAXHkfrfCn1plJG2Nt3Zj18vyY
CkU0BNDeRs3ajzWc+MRAUvBQMRlPotz3IXyiNM65iI2HW5UBLa0=
=uUi8
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Review by pinheadmz — APPROVED** [review]
ACK 9f97feee959adaf8c9319f85ae963ba4348f578f
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 9f97feee959adaf8c9319f85ae963ba4348f578f
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmKkweIACgkQ5+KYS2KJ
yTpwPRAAjgEsQJunnUycySVkvGdUoSo0ldekMdrTZPVRnc6M2S9MDKVniORoKQ/T
ojoK8BLjUqZNi+gNFkgVLQhGrqsmSERVbC3uAYXzP6DkWYqiMrONoOnlaGccQ0gc
x0bHxUvd3PT3x0Y2MucVQ/hQ3PkBX7B25PN3PuMb5LOuSc42dYwcO+Ml9HSE/DaB
2ALOpeF+tzCVhMEY2xQHzqzsRNuoKoc1Sm6eYs3azrdGC/VLJVpbNgUiSHcWNYZd
YYdWeLyt4SFfmWzOO2Q58VU4iZdhcPf6fDPHNd1waPunNSkRc5OnHlLdxrc47Acm
LeZdXwBT+xvjeBRmk6vkuEnzhmNvXaAVcaAPqbN3vCrF1/yxSInVxz56LpeUWFLQ
G6Fuee8hEAuG5IZhTQTzo00HgQJ5fky2rpk6WEDG0NtyLEr6RlUIKcBokZRaSAPs
GWY5w7Bkij23b4f3z4s/ApdL2fjzfbRd0LvExjp9RR7mx5wx1+h2YGQhD4eiYnYP
ELOMYJLmL3R4cPYNbDtRvcQtykgOr/VI5ro8zki15P8x4QkpCf/mQsPHM+Vm7ro+
gDeqiDygm+OJJliM4tt/cVeLeT19BtMEvodOS6oAIfEyJU8MZqiB5YOXLVuRjl9b
BxzRTw7bRWA1wyLRorPtcrvCQfzmGBL/bFwimGuUcobecxbJRfA=
=68Dn
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Comment by befranz** [comment]
> `--compact-tree-on-init` and `--compact-tree-init-interval` are similar to the `prune` option. So I would say, it depends on the configuration you are going for. For example, if you want Full Node in its truest sense - which needs to have fulll history of the tree and be able to provide proofs for the tree root in the history - then you don't use these options.

Thanks for your comprehensive answer. My script is basically for public listening full nodes, so this means those nodes can not benefit from the compaction, and a full tree history is needed to serve as such?

**Comment by pinheadmz** [comment]
IMO nodar is exaggerating with "full node in the truest sense." ;-)

All full nodes can safely compact the tree. The risk is that they won't survive a reorg greater than 250 blocks

**Comment by befranz** [comment]
> The risk is that they won't survive a reorg greater than 250 blocks

I see, we need at least a few nodes not compacted for the worst case. I guess easyhandshake is such a node. But which else, would it make sense to know/identify them?

**Comment by pinheadmz** [comment]
We considered adding a service flag to indicate limited tree history (we need to do this with pruned nodes as well) but a compacted tree is certainly enough to serve the network. 

In the event that a node operator wants to rebuild the tree completely this is actually available with an undocumented rpc reconstructtree and requires a full archival node.
